### ISSUE TITLE
How should we update version.php?

### ISSUE THREAD

**Issue by juho-jaakkola** [Elgg/Elgg#8463]
Traditionally the version number in ```version.php``` gets updated when we create a new upgrade script using ```engine/lib/upgrades/create_upgrade.php```.

I'm working on a new way of doing upgrades. Basically you create a class that contains all the logic and then you tell the system where to find that class when upgrades are being ran. This approach would make the ```create_upgrade.php``` obsolete.

Would it be ok, if we'd update ```version.php``` manually every time we add a new data upgrade? Or should we update the script to look for the new updates?

**Comment by mrclay** [comment]
What will we still need `$version` for?

**Comment by mrclay** [comment]
Should the upgrade classes announce (e.g. static method) a version number they're associated with so we can run them in a particular order?

**Comment by mrclay** [comment]
:arrow_up: I was thinking just core upgrades here

**Comment by ewinslow** [comment]
I'd love to remove any version numbers from the source code if possible.

**Comment by ewinslow** [comment]
Then the git tags become the source of truth and we don't have to duplicate that information. We could also delete the part of the script that bumps that number in composer.json

**Comment by juho-jaakkola** [comment]
> Should the upgrade classes announce (e.g. static method) a version number they're associated with so we can run them in a particular order?

Yes, my plan is that each upgrade has a date based version number ```yyyymmddxx```, and the upgrades will be ordered by date (until a more advanced dependency feature is added).

> What will we still need $version for?

Currently version number is used by:
 - ```create_upgrade.php``` to automatically increment the last two numbers if necessary
 - upgrading system to find out which upgrades have been ran
 - ```elgg_get_version()``` which is public and most likely used by many plugins

The new system could save the version number to database. That way ```elgg_get_version()``` could still work. I'm thinking that core and plugin upgrades would be ran separately, and the version number would be updated based on core upgrades, but not plugin upgrades.

**Comment by ewinslow** [comment]
So... I think `$version` needs to be burned with fire. Can we just kill it in 2.0 for good? Date-based versioning doesn't work at all for us since we support multiple parallel branches at once.

**Comment by mrclay** [comment]
We need an action plan for this, otherwise we should burn it with fire in 3.0.

**Comment by juho-jaakkola** [comment]
Not a blocker for 2.0
